strings/manacher_algorithm.cpp

```diff
@@ -0,0 +1,165 @@
+                    palindrome_max_half_length[opposite_to_i], right - i);
+            }
+
+            // expanding the palindrome across the maximum stored length in the
+            // array, centered at i
+            while (stuffed_string[i + (palindrome_max_half_length[i] + 1)] ==
+                   stuffed_string[i - (palindrome_max_half_length[i] + 1)]) {
+                palindrome_max_half_length[i]++;
+            }
+
+            // if palindrome centered at i exceeds the rightmost end of
+            // palindrome centered at 'bigger_center', then i will be made the
+            // 'bigger_center' and right value will also be updated with respect
+            // to center i
+            if (i + palindrome_max_half_length[i] > right) {
+                bigger_center = i;
+                right = i + palindrome_max_half_length[i];
+            }
+        }
+
+        // now extracting the first largest palindrome
+        int half_length = 0;   // half length of the largest palindrome
+        int center_index = 0;  // index of center of the largest palindrome
+
+        for (int i = 1; i < stuffed_string.size() - 1; i++) {
+            if (palindrome_max_half_length[i] > half_length) {
+                half_length = palindrome_max_half_length[i];
+                center_index = i;
+            }
+        }
+
+        std::string palindromic_substring =
+            "";  // contains the resulting largest palindrome
+
+        if (half_length > 0) {
+            // extra information: when '#' is the center, then palindromic
+            // substring will have even length, else palindromic substring will
+            // have odd length
+
+            int start = center_index - half_length +
+                        1;  // index of first character of palindromic substring
+            int end = center_index + half_length -
+                      1;  // index of last character of palindromic substring
+            for (int index = start; index <= end; index += 2) {
+                palindromic_substring += stuffed_string[index];
+            }
+        } else {
+            // if length = 0, then there does not exist any palindrome of length
+            // > 1 so we can assign any character of length 1 from string as the
+            // palindromic substring
+            palindromic_substring = prototype[0];
+        }
+        return palindromic_substring;
+
+    } else {
+        // handling case when string is empty
+        return "";
+    }
+}
+
+}  // namespace manacher
+}  // namespace strings
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    assert(strings::manacher::manacher("") == "");
+    assert(strings::manacher::manacher("abababc") == "ababa");
+    assert(strings::manacher::manacher("cbaabd") == "baab");
+    assert(strings::manacher::manacher("xacdedcax") == "xacdedcax");
+    assert(strings::manacher::manacher("xacddcax") == "xacddcax");
```

Comment:
Add some test cases with Upper , Mixed, camel Case. Test Cases format seem to be repeativie.